ww3tools/ProcSat_interpolation.py

```diff
@@ -0,0 +1,486 @@
+        max_lat_val = max_latitude.values.item()
+        min_lon_val = min_longitude.values.item()
+        max_lon_val = max_longitude.values.item()
+
+        within_lat_bounds = (satellite_latitudes >= min_lat_val) & (satellite_latitudes <= max_lat_val)
+        within_lon_bounds = (satellite_longitudes >= min_lon_val) & (satellite_longitudes <= max_lon_val)
+        return within_lat_bounds & within_lon_bounds
+
+    def prepare_and_interpolate(model_data, variable_name, satellite_latitudes, satellite_longitudes, satellite_times_unix, valid_times_unix):
+        mask = spatial_mask(satellite_latitudes, satellite_longitudes)
+        valid_satellite_points = np.array([satellite_latitudes[mask], satellite_longitudes[mask], satellite_times_unix[mask]]).T
+
+        variable_data = model_data[variable_name]
+        # Ensure the data is in the correct dimension order
+        if variable_data.dims != ('latitude', 'longitude', 'time'):
+            variable_data = variable_data.transpose('latitude', 'longitude', 'time')
+
+        model_grid = (variable_data.coords['latitude'].values,
+                      variable_data.coords['longitude'].values,
+                      valid_times_unix)
+
+        interpolator = RegularGridInterpolator(model_grid, variable_data.values, bounds_error=False, fill_value=np.nan)
+        interpolated_values = interpolator(valid_satellite_points)
+
+        # Handle model times for interpolated values based on satellite times
+        model_times_for_interpolated_values = np.searchsorted(valid_times_unix, satellite_times_unix[mask], side='right') - 1
+        model_times_for_interpolated_values = valid_times_unix[model_times_for_interpolated_values]
+
+        # Apply NaN where interpolated values are NaN
+        nan_indices = np.isnan(interpolated_values)
+        model_times_for_interpolated_values[nan_indices] = np.nan
+
+        full_interpolated_values = np.full(len(satellite_times_unix), np.nan)
+        full_interpolated_values[mask] = interpolated_values
+
+        full_model_times = np.full(len(satellite_times_unix), np.nan)
+        full_model_times[mask] = model_times_for_interpolated_values
+
+        return full_interpolated_values, full_model_times
+
+    # Interpolate for each variable
+    interpolated_htsgw_surface_values, interpolated_htsgw_surface_times = prepare_and_interpolate(
+        model_data_combined, 'HTSGW_surface', filtered_satellite_latitudes, filtered_satellite_longitudes, filtered_satellite_times_unix, model_time_seconds)
+
+    interpolated_wind_surface_values, interpolated_wind_surface_times = prepare_and_interpolate(
+        model_data_combined, 'WIND_surface', filtered_satellite_latitudes, filtered_satellite_longitudes, filtered_satellite_times_unix, model_time_seconds)
+    
+    fcst_hr = np.full_like(interpolated_wind_surface_values, np.nan, dtype='float64')
+    fcst_hr2_ws = (interpolated_wind_surface_times - model_reference_time_unix) / 3600
+    print(fcst_hr2_ws)
```

Comment:
@Ghazal-Mohammadpour this looks like a debugging print statement, should it remain here?